spin/test_helpers.go

```diff
@@ -0,0 +1,190 @@
+func (byPath TerraformModuleByPath) Less(i, j int) bool { return byPath[i].Path < byPath[j].Path }
+
+type RunningModuleByPath []*runningModule
+func (byPath RunningModuleByPath) Len() int           { return len(byPath) }
+func (byPath RunningModuleByPath) Swap(i, j int)      { byPath[i], byPath[j] = byPath[j], byPath[i] }
+func (byPath RunningModuleByPath) Less(i, j int) bool { return byPath[i].Module.Path < byPath[j].Module.Path }
+
+// We can't use assert.Equals on TerraformModule or any data structure that contains it because it contains some
+// fields (e.g. TerragruntOptions) that cannot be compared directly
+func assertModuleListsEqual(t *testing.T, expectedModules []*TerraformModule, actualModules []*TerraformModule, messageAndArgs ...interface{}) {
+	if !assert.Equal(t, len(expectedModules), len(actualModules), messageAndArgs...) {
+		t.Logf("%s != %s", expectedModules, actualModules)
+		return
+	}
+
+	sort.Sort(TerraformModuleByPath(expectedModules))
+	sort.Sort(TerraformModuleByPath(actualModules))
+
+	for i := 0; i < len(expectedModules); i++ {
+		expected := expectedModules[i]
+		actual := actualModules[i]
+		assertModulesEqual(t, expected, actual, messageAndArgs...)
+	}
+}
+
+// We can't use assert.Equals on TerraformModule because it contains some fields (e.g. TerragruntOptions) that cannot
+// be compared directly
+func assertModulesEqual(t *testing.T, expected *TerraformModule, actual *TerraformModule, messageAndArgs ...interface{}) {
+	if assert.NotNil(t, actual, messageAndArgs...) {
+		assert.Equal(t, expected.Config, actual.Config, messageAndArgs...)
+		assert.Equal(t, expected.Path, actual.Path, messageAndArgs...)
+
+		assertOptionsEqual(t, *expected.TerragruntOptions, *actual.TerragruntOptions, messageAndArgs...)
+		assertModuleListsEqual(t, expected.Dependencies, actual.Dependencies, messageAndArgs...)
+	}
+}
+
+// We can't use assert.Equals on TerraformModule or any data structure that contains it (e.g. runningModule) because it
+// contains some fields (e.g. TerragruntOptions) that cannot be compared directly
+func assertRunningModuleMapsEqual(t *testing.T, expectedModules map[string]*runningModule, actualModules map[string]*runningModule, doDeepCheck bool, messageAndArgs ...interface{}) {
+	if !assert.Equal(t, len(expectedModules), len(actualModules), messageAndArgs...) {
+		t.Logf("%s != %s", expectedModules, actualModules)
+		return
+	}
+
+	for expectedPath, expectedModule := range expectedModules {
+		actualModule, containsModule := actualModules[expectedPath]
+		if assert.True(t, containsModule, messageAndArgs...) {
+			assertRunningModulesEqual(t, expectedModule, actualModule, doDeepCheck, messageAndArgs...)
+		}
+	}
+}
+
+// We can't use assert.Equals on TerraformModule or any data structure that contains it (e.g. runningModule) because it
+// contains some fields (e.g. TerragruntOptions) that cannot be compared directly
```

Comment:
I know I say it a lot, but I spent *hours* in this PR working around limitations in Go, so I'll say it again: it's a shitty language.